pkt-meta/src/flow_table/thread_local_pq.rs

```diff
@@ -0,0 +1,261 @@
+    /// holding a lock on any element in the priority queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any lock acquired by this method is poisoned.
+    pub fn reap_all_expired(
+        &self,
+        on_expired: impl Fn(&Instant, &K, &V) -> PQAction,
+        on_reaped: impl Fn(K, V),
+    ) -> usize {
+        self.reap_all_expired_with_time_internal(&Instant::now(), &on_expired, &on_reaped)
+    }
+
+    #[allow(unused)] // This is unused for now if shuttle is not enabled
+    #[cfg(test)]
+    pub fn reap_all_expired_with_time(
+        &self,
+        now: &Instant,
+        on_expired: impl Fn(&Instant, &K, &V) -> PQAction,
+        on_reaped: impl Fn(K, V),
+    ) -> usize {
+        self.reap_all_expired_with_time_internal(now, &on_expired, &on_reaped)
+    }
+
+    fn reap_all_expired_with_time_internal(
+        &self,
+        now: &Instant,
+        on_expired: impl Fn(&Instant, &K, &V) -> PQAction,
+        on_reaped: impl Fn(K, V),
+    ) -> usize {
+        let pqs = self.pqs.iter();
+        let mut count = 0;
+        for pq in pqs {
+            let mut pq = pq.write().unwrap();
+            count += Self::reap_expired_locked_with_time(&mut pq, now, &on_expired, &on_reaped);
+        }
+        count
+    }
+
+    fn reap_expired_locked(
+        pq: &mut concurrency::sync::RwLockWriteGuard<
+            PriorityQueue<Entry<K, V>, Priority, RandomState>,
+        >,
+        on_expired: &impl Fn(&Instant, &K, &V) -> PQAction,
+        on_reaped: &impl Fn(K, V),
+    ) -> usize {
+        Self::reap_expired_locked_with_time(pq, &Instant::now(), on_expired, on_reaped)
+    }
+
+    fn reap_expired_locked_with_time(
+        pq: &mut concurrency::sync::RwLockWriteGuard<
+            PriorityQueue<Entry<K, V>, Priority, RandomState>,
+        >,
+        now: &Instant,
+        on_expired: &impl Fn(&Instant, &K, &V) -> PQAction,
+        on_reaped: &impl Fn(K, V),
+    ) -> usize {
+        let mut expired = Vec::new();
+        debug!(
+            "Reaping expired flows at {:?}, queue size {}",
+            now,
+            pq.len()
+        );
+        while let Some((_, expires_at)) = pq.peek() {
+            if *now >= expires_at.0 {
+                let ret = pq.pop();
+                let Some(entry) = ret else {
+                    break;
+                };
+                // This is going to copy the entry and key, even if it is to be reinserted,
+                // which sucks.  Find a better way to do this and placate the rust
+                // borrow checker.  Without this copy, the borrow checker will
+                // complain that you cannot pop the entry because of the borrow in the peek.
+                //
+                // This is probably fine for now though as we use K that is a FlowKey,
+                // copying it isn't ideal but probably cheap enough, and the value is an Arc.
+                expired.push(entry);
```

Comment:
Could you apply the action immediately here, and avoid `pop`-ping but simply call `change_priority()` if the caller asked to update the timestamp?

Reply:
No for complex borrow checker reasons that I can get in to if you;d like.  But in essence you need a mut and non-mut ref at the same time given the interface.